runtime/near-vm-logic/src/logic.rs

```diff
@@ -2454,28 +2454,35 @@ impl<'a> VMLogic<'a> {
             .into());
         }
         self.gas_counter.pay_per(storage_read_key_byte, key.len() as u64)?;
-        let nodes_before = self.ext.get_trie_nodes_count();
-        let read_mode = if cfg!(feature = "protocol_feature_flat_state") {
-            StorageGetMode::FlatStorage
+        let (value_ptr, tn_db_reads, tn_mem_reads) = if checked_feature!(
+            "protocol_feature_flat_state",
+            FlatStorageReads,
+            self.current_protocol_version
+        ) {
+            (self.ext.storage_get(&key, StorageGetMode::FlatStorage), 0, 0)
         } else {
-            StorageGetMode::Trie
+            let nodes_before = self.ext.get_trie_nodes_count();
+            let value_ptr = self.ext.storage_get(&key, StorageGetMode::Trie);
+            let nodes_delta = self
+                .ext
+                .get_trie_nodes_count()
+                .checked_sub(&nodes_before)
+                .ok_or(InconsistentStateError::IntegerOverflow)?;
+            self.gas_counter.add_trie_fees(&nodes_delta)?;
+            (value_ptr, nodes_delta.db_reads, nodes_delta.mem_reads)
```

Comment:
Sorry I am a bit confused about this change.

Why not keep the old code structure? It seemed more straight forward to me. 

Is the only reason to not call `self.ext.get_trie_nodes_count()`? Would that be a problem? 

Or are you worried about the TTN values not being zero if called in the state viewer? I don't think this actually fixes the state viewer, it will still fail on the next write because the chunk cache is wrong.

The only other case where I can see this matter is when calling this in FS mode but FS is not ready yet. In that case, we are already screwed and will not be able to produce correct results. (Maybe for the read we can but we will spoil the chunk cache and the next write call will therefore produce wrong TTN numbers.) 

In any case, if this change is for a specific cause, please describe it in a comment right next to the code.

Reply:
Good find. I wanted to make sure that TTN is completely eliminated from storage read costs, and possibly support state-viewer. But, as you said, chunk cache will be spoiled.

We need to either simplify change back, or come up with a way not to spoil chunk cache. And it deserves a special test.

Reply:
Hm, this could be ugly to implement, but maybe we can just somehow "turn off" the chunk cache for specific reads?

But also, do we need to solve this case? Or is it fine to e.g. `panic!("You can't run on protocol version X without flat storage")` for nodes that migrated too late? And for the state viewer, maybe it's fine to fix it later?

Reply:
Yeah, it's not necessary, we can fix it later. But another benefit, [which we discuss in Zulip at the same time](https://near.zulipchat.com/#narrow/stream/345766-pagoda.2Fstorage.2Fflat-storage/topic/Release.20plan/near/340891748), is to making FS optional for smoother release.

That's why I'm not sure about the right decision. In theory, we can call `set_trie_cache_mode(TrieCacheMode::CachingShard)` inside `storage_read` and switch it back afterwards, but I don't like it. Maybe we can come up with better solution.

Reply:
Rolled back to logic we had before, adding `checked_feature`. Now this change is trivial